README.md
# elasticsearch

[![Build Status](https://travis-ci.org/janis-commerce/elasticsearch.svg?branch=master)](https://travis-ci.org/janis-commerce/elasticsearch)
[![Coverage Status](https://coveralls.io/repos/github/janis-commerce/elasticsearch/badge.svg?branch=master)](https://coveralls.io/github/janis-commerce/elasticsearch?branch=master)

## Installation
```sh
npm install --save @janiscommerce/elasticsearch
```

## API

### `new ElasticSearch({config})`

Constructs the Elasticsearch driver instance, connected with the `config [Object]`.  

**Config definition (all settings are optional):**  
- protocol `[String]`: The protocol for the elasticsearch host url, not needed if you use a full url as host.  
- host `[String]`: Elasticsearch host URL. **In case of the host url includes a protocol, it will replace the actual protocol value in the config**.  
- port `[Number]`: The port for the elasticsearch host url, not needed if you use a full url as host.  
- user `[String]`: Elasticsearch user for the host.  
- password `[String]`: Elasticsearch password for the host.  
- limit `[Number]`: Default limit for getting operations.  
- prefix `[String]`: Prefix for table name, if it's setted, any operation will use the model table name with the prefix. Ex: `table.prefix`.  
- awsCredentials `[Boolean]`: Set to `true` if you need to use AWS credentials ENV variables for the host connection/authentication.  
- requestTimeout `[Number]`: The max timeout for operation in miliseconds, default 4000 (4s).  
- maxRetries `[Number]`: The max retries for operation, default 3.

**Config usage:**  
```js
{
	protocol: 'https://', // Default empty
	host: 'url-to-my-elasticsearch-host.com', // Default http://localhost
	port: 9200, // Default: empty
	user: 'Username', // Default empty
	password: 'password', // Default empty
	limit: 10, // Default 500
	prefix: 'products', // Default empty
	awsCredentials: true, // Default false
	requestTimeout: 2000, // Default 4000
	maxRetries: 1 // Default 3
}
```

### ***async*** `buildIndex(model)`

Puts the elasticsearch mappings obtained from the model, needed for sorting operations.  
**This method must be used once, only when a new table (elasticsearch index) is created.**  
**Is strongly recommended the usage of this method before any other operation.**

**Model example for this method:**

```js
class MyModel extends Model{

	static get table() {
		return 'table'; // The elasticsearch index
	}

	static get sortableFields() {
		return {
			id: { type: 'integer' }, // If there is no defined ID field, it will be mapped as text
			myfield: true, // Default will be mapped as text (elasticsearch string)
			myOtherField: { type: 'integer' },
			someOtherField: { type: 'long' }
		};
	}
}
```

**Elasticsearch basic field datatypes:**  

- **String:** text, keyword  
- **Numeric:** long, integer, short, byte, double, float, half_float, scaled_float  
- **Date:** date  
- **Date nanoseconds:** date_nanos  
- **Boolean:** boolean  
- **Binary:** binary  
- **Range:** integer_range, float_range, long_range, double_range, date_range  

**Important:**  
- If you define the `id` field as non text format may cause errors if you do an insertion operation without specifing the `id` field due the generated `id` will be in text format.  
- The table name from the model will be converted into snake_case due elasticsearch doesn't support camelCase in the index names. 
- The sortable fields are optional, same as this method, is only required for sorting items in get operations.

### ***async*** `insert(model, {item})`

Inserts an item into elasticsearch.  
Requires a `model [Model]` and `item [Object]`  
Returns the ID of the inserted item

**Important:**  

* If your item doesn't have an `id` field, it will be inserted with an auto-generated `id`, using `UUID`.  
* The `id` field is unique, it cannot be duplicated.  

### ***async*** `multiInsert(model, [{items}])`

Inserts multiple items into elasticsearch.  
Requires a `model [Model]` and `items [Object array]`  
Returns `true` if the operation was successful or `false` if not.

**Important:**  

* Every item that doesn't have an `id` field, will be inserted with an auto-generated `id` using `UUID`.  
* The `id` field is unique, it cannot be duplicated.  

### ***async*** `update(model, {values}, {filter})`

Updates one or multiple items from elasticsearch.  
Requires a `model [Model]`, `values [Object]` and `filter [Object]`  
Returns the updated count `[Number]`

### ***async*** `get(model, {params})`

Get items from the database then returns an `[Object array]` with the getted items.  
Requires a `model [Model]` and `params [Object]`  
Returns all items if you use without parameters  

Parameters (all are optional):  
* order `[Object]`: Order params for getted items, Example: `{ myField: 'asc', myOtherField: 'desc' }`  
* limit `[Number]`: Max amount of items per page to get. Default 500 or setted on config when constructs.  
* page `[Number]`: Get the items of the specified page.  
* filters `[Object]`: Search filters, leave empty for all items.  

**Filter operators**
- **$eq**: Matches values that are equal to a specified value.  
- **$gt**: Matches values that are greater than a specified value.  
- **$gte**: Matches values that are greater than or equal to a specified value.  
- **$in**: Matches any of the values specified in an array.  
- **$lt**: Matches values that are less than a specified value.  
- **$lte**: Matches values that are less than or equal to a specified value.  
- **$ne**: Matches all values that are not equal to a specified value.  
- **$nin**: Matches none of the values specified in an array.  

**Filters example**  

```js
{
	filters:{
		name: 'Edward', // Matches exactly the field, get only the items with 'Edward' in name field
		$in: {
			fullname: ['Edward', 'Sanchez'] // Similar as MongoDB, matches any of the words in the fullname field and, in this case, then gets all items that includes 'Edward' or 'Sanchez' in that field
		},
		$gt: {
			id: 10 // Similar as MongoDB, get all items which id field is greater than 10
		}
	}
}
```

### ***async*** `getTotals(model)`

Get the totals from the latest get operation with pagination.  
Requires a `model [Model]`  
Returns an `[Object]` with the total count, page size, total pages and current page.  

**getTotals return example**  

```js
{
	total: 1000,
	pageSize: 100, // Limit from the last get operation or default value (500)
	pages: 10,
	page: 5
}
```

### ***async*** `save(model, {item})`

Insert/update an item into the elasticsearch.  
Requires a `model [Model]` and `item [Object]`  
Returns the ID of the upserted item  

### ***async*** `multiSave(model, [{items}])`

Insert/update multiple items into the elasticsearch.  
Requires a `model [Model]` and `items [Object array]`  
Returns `true` if the operation was successful or `false` if not.  

### ***async*** `remove(model, {filters})`

Removes the item or items that matches the filters from the elasticsearch.  
Requires a `model [Model]` and `filters [Object]`  
Returns the deleted count `[Number]`  

### ***async*** `distinct(model, {params})`

Requires a `model [Model]` and `params [Object]`  
Returns an `[String array]` with the unique values of the key field from elasticsearch.  

Parameters:  
* key `[String]` (required): The key field name  
* filters `[Object]` (optional): Search filters, leave empty for all items.  

## Errors

The errors are informed with a `ElasticSearchError`.
This object has a code that can be useful for a correct error handling.  
The codes are the following:  

| Code | Description                    |
|------|--------------------------------|
| 1    | Invalid model                  |
| 2    | Invalid query                  |
| 3    | Internal elasticsearch error   |
| 4    | Invalid filters                |
| 5    | Invalid filter operator        |
| 6    | Index not found                |
| 7    | Index not built                |
| 8    | Invalid distinct key           |

The config validation errors are informed with a `ElasticSearchConfigError`.
This object has a code that can be useful for a correct error handling.  
The codes are the following:  

| Code | Description                    |
|------|--------------------------------|
| 1    | Invalid config                 |
| 2    | Invalid setting                |

## Usage
```js
const ElasticSearch = require('@janiscommerce/elasticsearch');
const Model = require('myModel');

const elastic = new ElasticSearch({
	host: 'https://myelastichost.com',
	user: 'root',
	password: 'foobar'
});

const model = new Model();

(async () => {

	await elastic.buildIndex(model);

	let result;

	// insert
	result = await elastic.insert(model, { id: 1, value 'sarasa' }); // expected return: 1

	// multiInsert
	result = await elastic.multiInsert(model, [
		{ id: 1, value: 'sarasa 1' },
		{ id: 2, value: 'sarasa 2' },
		{ id: 3, value: 'sarasa 3' }
	]); // expected return: true

	// update
	result = await elastic.update(model, { value: 'foobar 1' }, { id: 1 }); // expected return: 1

	// get
	result = await elastic.get(model, {}); // expected return: all entries array [{item}, {item}...]
	result = await elastic.get(model, { filters: { id: 1 } }); // expected return: an array with the item with id 1
	result = await elastic.get(model, { limit: 10, page: 2, filters: { value: 'foo' } }); // expected return: the page 2 of the elements with value "foo" with a page size of 10.
	result = await elastic.get(model, { order: { id: 'asc' } }); // expected return: all entries array ordered ascendently by id

	// getTotals
	result = await elastic.getTotals(model);

	/* expected return:
	{
		page: 2,
		limit: 10,
		pages: 5,
		total: 50
	}
	*/

	// save
	result = await elastic.save(model, { id: 1, value: 'foobar 1' }); // example return: 1

	// multiSave
	result = await elastic.multiSave(model, [
		{ id: 1, value: 'foobar 1' },
		{ id: 2, value: 'foobar 2' },
		{ id: 3, value: 'foobar 3' }
	]); // expected return: true

	// remove
	result = await elastic.remove(model, { 
		$in: {
			value: ['foobar']
		} 
	}); // expected return: 3

	// distinct
	result = await elastic.distinct(model, {
		key: 'myField';
	}); // expected return: ['my-value', 'my-other-value']
})

```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.2.2] - 2019-12-06
### Added
- `requestTimeout` and `maxRetries` support in config

## [1.2.1] - 2019-12-03
### Fixed
- `distinct` will return all posible results instead 10

## [1.2.0] - 2019-12-02
### Added
- `distinct` method

### Fixed
- `ElasticSearchFilters` field datatypes will be filtered by `raw` only when exists in the model `sortableFields` otherwise will use `keyword` instead.

### Removed
- unnecessary `index.js` files

## [1.1.0] - 2019-09-12
### Added
- `ElasticSearchConfigError`
- `ConfigValidator` module for config datatypes validation
- `prefix` support for table names

## [1.0.2] - 2019-08-29
### Fixed
- `getTotals` params obtained from latest `get` operation
- Error handling improved, and clear messages added for most common ElasticSearch errors

## [1.0.1] - 2019-08-28
### Fixed
- AWS Authentication
- `lastModified` field name changed to `dateModified`

## [1.0.0] - 2019-08-27
### Added
- ElasticSearch package
- Unit tests
- Docs

tests/config-validator-test.js
'use strict';

const assert = require('assert');

const ConfigValidator = require('../lib/config-validator');

const ElasticSearchConfigError = require('../lib/elasticsearch-config-error');

describe('ConfigValidator', () => {

	describe('validate()', () => {

		it('should throw invalid config when the config is empty', () => {

			assert.throws(() => ConfigValidator.validate(), {
				name: 'ElasticSearchConfigError',
				code: ElasticSearchConfigError.codes.INVALID_CONFIG
			});

		});


		it('should throw invalid config when the config is not an object', () => {

			assert.throws(() => ConfigValidator.validate('string'), {
				name: 'ElasticSearchConfigError',
				code: ElasticSearchConfigError.codes.INVALID_CONFIG
			});

		});


		it('should throw invalid config when the config is an array', () => {

			assert.throws(() => ConfigValidator.validate([]), {
				name: 'ElasticSearchConfigError',
				code: ElasticSearchConfigError.codes.INVALID_CONFIG
			});

		});


		['string', ['array']].forEach(type => {

			it('should throw invalid setting when a setting has an unexpected type', () => {

				assert.throws(() => ConfigValidator.validate({
					prefix: 'myPrefix',
					port: type
				}), {
					name: 'ElasticSearchConfigError',
					code: ElasticSearchConfigError.codes.INVALID_SETTING
				});
			});

		});

		it('should not throw when the settings are correct', () => {

			assert.doesNotThrow(() => ConfigValidator.validate({
				prefix: 'myPrefix',
				port: 9200
			}));
		});
	});
});


tests/elasticsearch-filters-test.js
'use strict';

const assert = require('assert');

const ElasticSearchFilters = require('../lib/elasticsearch-filters');

const ElasticSearchError = require('../lib/elasticsearch-error');

class Model {

	static get table() {
		return 'myTable';
	}

	static get sortableFields() {
		return {
			id: true,
			value: true,
			othervalue: { type: 'integer' }
		};
	}
}

const model = new Model();

describe('ElasticSearchFilters', () => {

	describe('getFilters()', () => {

		it('should return the elasticsearch query (term only)', () => {

			const termFilters = {
				id: 'value',
				$eq: {
					id: 'value'
				},
				$ne: {
					field: 'value'
				},
				$in: {
					field: ['value']
				},
				$nin: {
					field: ['value']
				}
			};

			assert.deepStrictEqual(ElasticSearchFilters.getFilters(model, termFilters), {
				query: {
					bool: {
						must: [
							{ term: { 'id.raw': 'value' } },	{ terms: { field: ['value'] } }
						],
						must_not: [
							{ term: { 'field.keyword': 'value' } }, { terms: { field: ['value'] } }
						]
					}
				}
			});
		});

		it('should return the elasticsearch query (range only)', () => {

			const rangeFilters = {
				$gt: {
					id: 1
				},
				$gte: {
					id: 1
				},
				$lt: {
					id: 1
				},
				$lte: {
					id: 1
				}
			};

			assert.deepStrictEqual(ElasticSearchFilters.getFilters(model, rangeFilters), {
				query: {
					range: {
						id: {
							gt: 1,
							gte: 1,
							lt: 1,
							lte: 1
						}
					}
				}
			});
		});

		it('should return the elasticsearch query (mixed)', () => {

			const mixedFilters = {
				$ne: {
					field: 'value'
				},
				$nin: {
					field: ['value']
				},
				$gt: {
					id: 1
				},
				$gte: {
					id: 1
				},
				$lt: {
					id: 1
				},
				$lte: {
					id: 1
				}
			};

			assert.deepStrictEqual(ElasticSearchFilters.getFilters(model, mixedFilters), {
				query: {
					bool: {
						must: [
							{
								range: {
									id: {
										gt: 1,
										gte: 1,
										lt: 1,
										lte: 1
									}
								}
							}
						],
						must_not: [
							{ term: { 'field.keyword': 'value' } }, { terms: { field: ['value'] } }
						]
					}
				}
			});
		});

		it('should throw \'Invalid filters\' when the $in terms values aren\'t an array', async () => {

			assert.throws(() => ElasticSearchFilters.getFilters(model, {
				$in: 'not-array'
			}), {
				name: 'ElasticSearchError',
				code: ElasticSearchError.codes.INVALID_FILTERS
			});

		});

		it('should throw \'Invalid filters\' when the $nin terms values aren\'t an array', async () => {

			assert.throws(() => ElasticSearchFilters.getFilters(model, {
				$nin: 'not-array'
			}), {
				name: 'ElasticSearchError',
				code: ElasticSearchError.codes.INVALID_FILTERS
			});

		});

		context('when the recieved filters are invalid', () => {

			it('should throw \'Invalid filter operator\' when any of the received operator not exists', async () => {

				assert.throws(() => ElasticSearchFilters.getFilters(model, {
					$eq: { field: 'value' },
					$eqq: { field: 'value' }
				}), {
					name: 'ElasticSearchError',
					code: ElasticSearchError.codes.INVALID_FILTER_OPERATOR
				});

			});

			it('should throw \'Invalid filters\' when the filters not exists', async () => {
				assert.throws(() => ElasticSearchFilters.getFilters(model), {
					name: 'ElasticSearchError',
					code: ElasticSearchError.codes.INVALID_FILTERS
				});
			});

			it('should throw \'Invalid filters\' when the filters is not an object', async () => {
				assert.throws(() => ElasticSearchFilters.getFilters(model, 'filters'), {
					name: 'ElasticSearchError',
					code: ElasticSearchError.codes.INVALID_FILTERS
				});
			});

			it('should throw \'Invalid filters\' when the filters is an array', async () => {
				assert.throws(() => ElasticSearchFilters.getFilters(model, ['filters']), {
					name: 'ElasticSearchError',
					code: ElasticSearchError.codes.INVALID_FILTERS
				});
			});
		});

		context('when the parent properties of the parsed filters not exists', () => {

			/* eslint-disable no-underscore-dangle */

			[
				ElasticSearchFilters._formatEq.bind(ElasticSearchFilters),
				ElasticSearchFilters._formatIn

			].forEach(formatter => {

				it('should create the bool property when not exists', () => {

					const parsedFilters = {};

					formatter(parsedFilters, 'field', ['value']);

					assert.strictEqual(typeof parsedFilters.bool, 'object');
				});

				it('should create the bool.must property when not exists', () => {

					const parsedFilters = {};

					formatter(parsedFilters, 'field', ['value']);

					assert.strictEqual(Array.isArray(parsedFilters.bool.must), true);
				});
			});

			[
				ElasticSearchFilters._formatNe.bind(ElasticSearchFilters),
				ElasticSearchFilters._formatNin

			].forEach(formatter => {

				it('should create the bool property when not exists', () => {

					const parsedFilters = {};

					formatter(parsedFilters, 'field', ['value']);

					assert.strictEqual(typeof parsedFilters.bool, 'object');
				});

				it('should create the bool.must_not property when not exists', () => {

					const parsedFilters = {};

					formatter(parsedFilters, 'field', ['value']);

					assert.strictEqual(Array.isArray(parsedFilters.bool.must_not), true);
				});
			});

			[
				ElasticSearchFilters._formatGt,
				ElasticSearchFilters._formatGte,
				ElasticSearchFilters._formatLt,
				ElasticSearchFilters._formatLte

			].forEach(formatter => {

				it('should create the range property when not exists', () => {

					const parsedFilters = {};

					formatter(parsedFilters, 'field', 'value');

					assert(typeof parsedFilters.range === 'object');
				});

				it('should create the range[field] property when not exists', async () => {

					const parsedFilters = {};

					formatter(parsedFilters, 'field', 'value');

					assert(typeof parsedFilters.range.field === 'object');

				});

			});

			/* eslint-enable no-underscore-dangle */

		});
	});
});


lib/elasticsearch.js
'use strict';

const AWS = require('aws-sdk');

const AWSESConnection = require('http-aws-es');

const { URL } = require('url');

const UUID = require('uuid/v4');

const elasticsearch = require('elasticsearch');

const ElasticSearchFilters = require('./elasticsearch-filters');
const ElasticSearchError = require('./elasticsearch-error');

const ConfigValidator = require('./config-validator');

const DEFAULT_LIMIT = 500;
const DEFAULT_TIMEOUT = 4000;
const DEFAULT_RETRIES = 3;
const ES_DEFAULT_HOST = 'http://localhost';
const ES_MAX_BUCKET_SIZE = 2147483647;

function convertToSnakeCase(str) {
	return str.replace(/[A-Z]/g, letter => `_${letter.toLowerCase()}`);
}

/**
 * @class ElasticSearch
 * @classdesc Elasticsearch driver module
*/
class ElasticSearch {

	constructor(config = {}) {

		ConfigValidator.validate(config);

		this.config = {
			protocol: config.protocol || '',
			host: config.host || ES_DEFAULT_HOST,
			port: config.port || '',
			user: config.user || '',
			password: config.password || '',
			limit: config.limit || DEFAULT_LIMIT,
			prefix: config.prefix || '',
			awsCredentials: config.awsCredentials || false,
			requestTimeout: config.requestTimeout || DEFAULT_TIMEOUT,
			maxRetries: config.maxRetries || DEFAULT_RETRIES
		};

		try {

			const url = new URL(this.config.host); // throws if the URL doesn't have a protocol
			this.config.protocol = `${url.protocol}//`;
			this.config.host = url.host;

		} catch(err) {
			// do nothing...
		}
	}

	/**
	 * AWS Credentials config for Elasticsearch
	 * @returns {Object} AWS connection config
	 */
	get _awsCredentials() {

		if(!this.config.awsCredentials)
			return {};

		AWS.config.update({
			credentials: new AWS.EnvironmentCredentials('AWS'),
			region: process.env.AWS_DEFAULT_REGION
		});

		return {
			connectionClass: AWSESConnection
		};
	}

	/**
	 * Elasticsearch connection URL prefix
	 * @returns {String} Elasticsearch URL prefix
	*/
	get _userPrefix() {
		return this.config.user ? `${this.config.user}:${this.config.password}@` : '';
	}

	/**
	 * Checks the connection to the Elasticsearch Client
	 * @returns {elasticsearch.Client} Elasticsearch client with the connection
	 * @throws if the connection is not successfull
	*/
	get client() {

		if(!this._client) {
			try {
				this._client = new elasticsearch.Client({
					host: `${this.config.protocol}${this._userPrefix}${this.config.host}${this.config.port ? `:${this.config.port}` : ''}`,
					requestTimeout: this.config.requestTimeout,
					maxRetries: this.config.maxRetries,
					...this._awsCredentials
				});
			} catch(err) {
				this._handleElasticSearchError(err);
			}
		}

		return this._client;
	}

	/**
	 * Validates the model structure
	 * @param {Model} model Model instance
	 * @throws If the model is invalid
	 */
	_validateModel(model) {

		if(!model)
			throw new ElasticSearchError('Empty model', ElasticSearchError.codes.INVALID_MODEL);

		if(!model.constructor.table)
			throw new ElasticSearchError('Invalid model', ElasticSearchError.codes.INVALID_MODEL);

		if(model.constructor.sortableFields && (typeof model.constructor.sortableFields !== 'object' || Array.isArray(model.constructor.sortableFields))) {
			throw new ElasticSearchError('Invalid sortable fields, it must be an object. See model.constructor.sortableFields',
				ElasticSearchError.codes.INVALID_MODEL);
		}

	}

	/**
	 * Get the index name converted into snake_case is case of a prefix is setted it returns the index name with the prefix, concatenated with a dot.
	 * @param {Model} model Model instance
	 * @returns {String} Index/table name from the model, converted into snake_case
	 */
	_getIndex(model) {
		return convertToSnakeCase(
			this.config.prefix ? `${model.constructor.table}.${this.config.prefix}` : model.constructor.table
		);
	}

	/**
	 * Make more readable errors from the received ElasticSearch Error
	 * @param {Error} err Catched ElasticSearch Error
	 * @throws {Error} More clear and readable error message
	 */
	_handleElasticSearchError(err) {

		if(!(err.body && err.body.error))
			throw new ElasticSearchError(err.message, ElasticSearchError.codes.ELASTICSEARCH_ERROR);

		switch(err.body.error.type) {

			case 'version_conflict_engine_exception':
				throw new ElasticSearchError('Duplicated ID field',
					ElasticSearchError.codes.INVALID_QUERY);

			case 'mapper_parsing_exception':
				throw new ElasticSearchError(`Received field datatypes that not match with existing mappings: ${err.body.error.reason}`,
					ElasticSearchError.codes.INVALID_QUERY);

			case 'parsing_exception':
				throw new ElasticSearchError(`An error occurred while processing the query filters or sortings: ${err.body.error.reason}`,
					ElasticSearchError.codes.INVALID_QUERY);

			case 'search_phase_execution_exception':
				throw new ElasticSearchError(`The received filters not match with the field datatype or are invalid: ${err.body.error.reason}`,
					ElasticSearchError.codes.INVALID_QUERY);

			case 'index_not_found_exception':
				throw new ElasticSearchError('Index not found. See model.constructor.table or create it using buildIndex(model)',
					ElasticSearchError.codes.INDEX_NOT_FOUND);

			case 'query_shard_exception':
				throw new ElasticSearchError('Can\'t do sorting operations without mappings, create it using buildIndex(model) first',
					ElasticSearchError.codes.INDEX_NOT_BUILT);

			case 'illegal_argument_exception':
				throw new ElasticSearchError(`Can't modify already existing mappings datatypes: ${err.body.error.reason}. See model.constructor.sortableFields`,
					ElasticSearchError.codes.INVALID_MODEL);

			default:
				throw new ElasticSearchError(err.message, ElasticSearchError.codes.ELASTICSEARCH_ERROR);
		}

	}

	/**
	 * Build elasticsearch index with mappings for sortings
	 * @param {Model} model Model instance
	 */
	async buildIndex(model) {

		if(!model.constructor.sortableFields)
			return;

		this._validateModel(model);

		try {

			const exists = await this.client.indices.exists({ index: this._getIndex(model) });

			if(!exists)
				await this.client.indices.create({ index: this._getIndex(model) });

			await this.client.indices.putMapping({
				index: this._getIndex(model),
				body: this._getMappingsFromModel(model)
			});

		} catch(err) {
			this._handleElasticSearchError(err);
		}
	}

	/**
	 * Analyses the model fields and build the mappings query for elasticsearch
	 * @param {Model} model Model instance
	 * @returns {Object} elasticsearch mappings query
	 */
	_getMappingsFromModel(model) {

		const properties = {};

		const sortableFields = { ...model.constructor.sortableFields };

		if(!sortableFields.id)
			sortableFields.id = true;

		if(!sortableFields.dateCreated)
			sortableFields.dateCreated = { type: 'date' };

		if(!sortableFields.dateModified)
			sortableFields.dateModified = { type: 'date' };

		const fields = Object.keys(sortableFields);

		for(const field of fields) {

			const fieldType = sortableFields[field].type || 'text';

			properties[field] = {
				type: fieldType,
				fields: {
					raw: {
						type: fieldType === 'text' ? 'keyword' : fieldType
					}
				}
			};
		}

		return { properties };
	}

	/**
	 * Prepare the items for elasticsearch operations
	 * @param {Array} items items
	 * @returns {Array|Object} item/items with prepared fields
	 */
	_prepareFields(items) {

		if(!Array.isArray(items))
			items = [items];

		for(const item of items) {

			if(!item.id)
				item.id = UUID();

		}

		if(items.length === 1)
			return items[0];
		return items;
	}

	/**
	 * Parse the item into elasticsearch script string for updates
	 * @param {Object} item item
	 * @returns {Object} parsed item for updates
	 */
	_parseValuesForUpdate(item) {

		let parsedItem = '';

		Object.entries(item).forEach(([key, value]) => {

			switch(typeof value) {

				case 'string':
					parsedItem += `ctx._source.${key}='${value}';`;
					break;

				case 'object':
					parsedItem += `ctx._source.${key}=${JSON.stringify(value)};`;
					break;

				default:
					parsedItem += `ctx._source.${key}=${value};`;
			}

		});

		parsedItem += `ctx._source.dateModified='${new Date().toISOString()}';`;

		return { script: parsedItem };
	}

	/**
	 * Converts the items array into a multi insert bulk query for elasticsearch
	 * @param {Array} items items
	 * @returns {Array} parsed items
	 */
	_parseItemsForBulkInsert(items) {

		if(!Array.isArray(items))
			items = [items];

		const parsedItems = [];

		for(const item of items) {

			parsedItems.push(
				{ index: { _id: item.id } },
				{
					...item,
					...!item.dateCreated ? { dateCreated: new Date() } : {}
				}
			);

		}

		return parsedItems;
	}

	/**
	 * Converts the items array into a multi upsert bulk query for elasticsearch
	 * @param {Array} items items
	 * @returns {Array} parsed items
	 */
	_parseItemsForBulkUpsert(items) {

		if(!Array.isArray(items))
			items = [items];

		const parsedItems = [];

		for(const item of items) {

			parsedItems.push(
				{ update: { _id: item.id, _type: '_doc' } },
				{
					doc: { ...item, dateModified: new Date() },
					upsert: {
						...item,
						...!item.dateCreated ? { dateCreated: new Date() } : {}
					}
				}
			);

		}

		return parsedItems;
	}

	/**
	 * Parse the order params into a sort query for elasticsearch
	 * @param {Object} order order params
	 * @returns {Object} parsed sort query
	 */
	_parseSortingParams(order) {

		const parsedSorting = Object.entries(order).map(([key, value]) => {

			return {
				[`${key}.raw`]: value
			};

		});

		return { sort: parsedSorting };
	}

	/**
	 * Inserts data into elasticsearch
	 * @param {Model} model Model instance
	 * @param {Object} item item
	 * @returns {Boolean} true/false
	 */
	async insert(model, item) {

		this._validateModel(model);
		item = this._prepareFields(item);


		if(!item.dateCreated)
			item.dateCreated = new Date();

		try {
			const res = await this.client.create({
				index: this._getIndex(model),
				type: '_doc',
				id: item.id,
				refresh: 'wait_for',
				body: item
			});

			return res._id; // eslint-disable-line

		} catch(err) {
			this._handleElasticSearchError(err);
		}
	}

	/**
	 * Inserts multiple items into elasticsearch
	 * @param {Model} model Model instance
	 * @param {Array} items items
	 * @returns {Boolean} true/false
	 */
	async multiInsert(model, items) {

		this._validateModel(model);
		items = this._prepareFields(items);

		try {
			const res = await this.client.bulk({
				index: this._getIndex(model),
				type: '_doc',
				refresh: 'wait_for',
				body: this._parseItemsForBulkInsert(items)
			});

			return !res.errors;

		} catch(err) {
			this._handleElasticSearchError(err);
		}
	}

	/**
	 * Get data from elasticsearch
	 * @param {Model} model Model instance
	 * @param {Object} params parameters (limit, page, filters...)
	 * @returns {Array} elasticsearch result
	 */
	async get(model, params = {}) {

		this._validateModel(model);

		const limit = params.limit || this.config.limit;
		const page = params.page || 1;
		const filters = params.filters ? ElasticSearchFilters.getFilters(model, params.filters) : {};
		const order = params.order ? this._parseSortingParams(params.order) : {};

		try {
			const res = await this.client.search({
				index: this._getIndex(model),
				type: '_doc',
				body: {
					...filters,
					...order
				},
				from: limit * (page - 1),
				size: limit
			});

			model.lastQueryEmpty = !res.hits.total;
			model.totalsParams = { limit, page, filters, order };

			return res.hits.hits.map(item => item._source); // eslint-disable-line

		} catch(err) {
			this._handleElasticSearchError(err);
		}
	}

	/**
	 * Get the paginated totals from latest get query.
	 * @param {Model} model Model instance
	 * @returns {Object} total, page size, pages and page from the results.
	 */
	async getTotals(model) {

		this._validateModel(model);

		if(model.lastQueryEmpty)
			return { total: 0, pages: 0 };

		const params = model.totalsParams || {};

		const filters = params.filters ? params.filters : {};

		try {
			const res = await this.client.count({
				index: this._getIndex(model),
				type: '_doc',
				body: filters
			});

			const result = {
				total: res.count,
				pageSize: params.limit || this.config.limit,
				pages: params.limit ? Math.ceil(res.count / params.limit) : 1,
				page: params.page || 1
			};

			if(result.page > result.pages)
				result.page = result.pages;

			return result;

		} catch(err) {
			this._handleElasticSearchError(err);
		}
	}

	/**
	 * Updates data into the database
	 * @param {Model} model Model instance
	 * @param {Object} values values to apply
	 * @param {Object} filter filters
	 * @returns {Number} modified count
	 */
	async update(model, values, filters) {

		this._validateModel(model);

		try {
			const res = await this.client.updateByQuery({
				index: this._getIndex(model),
				refresh: true,
				type: '_doc',
				body: {
					...ElasticSearchFilters.getFilters(model, filters),
					...this._parseValuesForUpdate(values)
				}
			});

			return res.updated;

		} catch(err) {
			this._handleElasticSearchError(err);
		}
	}

	/**
	 * Insert/update one element into the database
	 * @param {Model} model Model instance
	 * @param {Object} item item
	 * @returns {Boolean} true/false
	 */
	async save(model, item) {

		this._validateModel(model);
		item = this._prepareFields(item);

		try {
			const res = await this.client.update({
				index: this._getIndex(model),
				type: '_doc',
				id: item.id,
				refresh: 'wait_for',
				body: {
					doc: { ...item, dateModified: new Date() },
					upsert: {
						...item,
						...!item.dateCreated ? { dateCreated: new Date() } : {}
					}
				}

			});

			return res._id; // eslint-disable-line

		} catch(err) {
			this._handleElasticSearchError(err);
		}
	}

	/**
	 * Multi insert/update items into the dabatase
	 * @param {Model} model Model instance
	 * @param {Array} items items
	 * @param {Number} limit specifies the limit of items that can be bulk writed into monogdb at the same time
	 * @returns {Boolean} true/false
	 */
	async multiSave(model, items) {

		this._validateModel(model);
		items = this._prepareFields(items);

		try {
			const res = await this.client.bulk({
				index: this._getIndex(model),
				refresh: 'wait_for',
				body: this._parseItemsForBulkUpsert(items)
			});

			return !res.errors;

		} catch(err) {
			this._handleElasticSearchError(err);
		}
	}

	/**
	 * Removes the item/items that match the filters
	 * @param {Model} model Model instance
	 * @param {Object} filters filters
	 * @returns {Number} deleted count
	 */
	async remove(model, filters) {

		this._validateModel(model);

		try {
			const res = await this.client.deleteByQuery({
				index: this._getIndex(model),
				type: '_doc',
				refresh: true,
				body: {
					...ElasticSearchFilters.getFilters(model, filters)
				}
			});

			return res.deleted;

		} catch(err) {
			this._handleElasticSearchError(err);
		}
	}

	/**
	 * Get distinct of a field
	 * @param {Model} model model instance
	 * @param {Object} params parameters (key and filters)
	 * @returns {Array} elasticsearch results
	 */
	async distinct(model, params = {}) {

		this._validateModel(model);

		const { key } = params;

		if(typeof key !== 'string')
			throw new ElasticSearchError('Invalid distinct key: Should exist and must be a string.', ElasticSearchError.codes.INVALID_DISTINCT_KEY);

		const filters = params.filters ? ElasticSearchFilters.getFilters(model, params.filters) : {};

		try {

			const res = await this.client.search({
				index: this._getIndex(model),
				type: '_doc',
				body: {
					...filters,
					aggs: {
						distinct: {
							terms: {
								field: `${key}.${ElasticSearchFilters.getFieldType(model, key)}`,
								size: ES_MAX_BUCKET_SIZE
							}
						}
					}
				}
			});

			return res.aggregations.distinct.buckets.map(bucket => bucket.key);

		} catch(err) {
			this._handleElasticSearchError(err);
		}
	}
}

module.exports = ElasticSearch;
